java-checks/src/main/java/org/sonar/java/checks/RedundantTypeCastCheck.java

```diff
@@ -62,89 +66,82 @@ public void scanFile(JavaFileScannerContext context) {
+      reportIssue(typeCastTree.type(), "Remove this unnecessary cast to \"" + cast + "\".");
     }
   }
 
-  private static boolean isChainedCastWithWildcard(TypeCastTree typeCastTree) {
-    ExpressionTree expression = typeCastTree.expression();
-    return expression.is(Tree.Kind.TYPE_CAST) && usesWildCard(expression);
-  }
-
-  private void addArgsToExclusion(Tree tree) {
-    List<ExpressionTree> args;
-    if(tree.is(Tree.Kind.METHOD_INVOCATION)) {
-      MethodInvocationTree mit = (MethodInvocationTree) tree;
-      args = mit.arguments();
-    } else {
-      NewClassTree newClassTree = (NewClassTree) tree;
-      args = newClassTree.arguments();
-    }
-    for (ExpressionTree arg : args) {
-      if (arg.is(Tree.Kind.TYPE_CAST)) {
-        excluded.add(arg);
+  @CheckForNull
+  private Type targetType(Tree tree) {
+    Tree parent = skipParentheses(tree.parent());
+    Type target = null;
+    if(parent.is(Tree.Kind.RETURN_STATEMENT)) {
+      Tree method = parent;
+      while (!method.is(Tree.Kind.METHOD)) {
+        method = method.parent();
+      }
+      target = ((JavaType.MethodJavaType) ((MethodTree) method).symbol().type()).resultType();
+    } else if (parent.is(Tree.Kind.VARIABLE)) {
+      VariableTree variableTree = (VariableTree) parent;
+      target = variableTree.symbol().type();
+    } else if (parent.is(Tree.Kind.METHOD_INVOCATION)) {
+      MethodInvocationTree mit = (MethodInvocationTree) parent;
+      if(mit.symbol().isMethodSymbol()) {
+        JavaSymbol.MethodJavaSymbol sym = (JavaSymbol.MethodJavaSymbol) mit.symbol();
+        int castArgIndex = mit.arguments().indexOf(tree);
+        target = sym.parameterTypes().get(castArgIndex);
       }
+    } else if(parent.is(Tree.Kind.MEMBER_SELECT)) {
+      MemberSelectExpressionTree mse = (MemberSelectExpressionTree) parent;
+      target = mse.identifier().symbol().owner().type();
+    } else if(parent instanceof ExpressionTree) {
+      target = ((ExpressionTree) parent).symbolType();
     }
+    return target;
   }
 
-  private static boolean isExcluded(JavaType cast) {
-    return cast.isUnknown();
+  private static Tree skipParentheses(Tree parent) {
+    while (parent.is(Tree.Kind.PARENTHESIZED_EXPRESSION)) {
+      parent = parent.parent();
+    }
+    return parent;
   }
 
-  private static boolean isRedundantCast(JavaType cast, JavaType expressionType) {
-    JavaType erasedExpressionType = expressionType;
-    if(erasedExpressionType.isTagged(JavaType.TYPEVAR)) {
-      erasedExpressionType = erasedExpressionType.erasure();
-    }
-    if (isParametrizedType(cast) ^ isParametrizedType(expressionType)) {
+  private static boolean isSubtype(Type expression, Type target) {
+    JavaType supType = (JavaType) target;
+    JavaType expType = (JavaType) expression;
+    if(expType.isTagged(JavaType.TYPEVAR) && supType.isTagged(JavaType.TYPEVAR)) {
+      if(supType == expression) {
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Change this comparison to use the equals method. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1698)